### PR TITLE
refactor CheckoutState

### PR DESCRIPTION
**Pull request by tomerd**
motivation: when working on configuration, CheckoutState data model stood out as one that can be improved to be safer

changes:
* transition CheckoutState from struct with optionals to enum representing the finite states
* adjust call sites and tests

### CODE REVIEW

File: Sources/Workspace/ResolverPrecomputationProvider.swift [hunk 1]
```diff
@@ -121,23 +122,25 @@ private struct LocalPackageContainer: PackageContainer {
...
         // Because of the implementation of `reversedVersions`, we should only get the exact same version.
-        precondition(dependency?.checkoutState?.version == version)
+        guard case .version(version, revision: _) = dependency?.checkoutState else {
+            throw InternalError("expected version, but checkout state was \(String(describing: dependency?.checkoutState))")
+        }
         return try manifest.dependencyConstraints(productFilter: productFilter)
     }
 
-    func getDependencies(at revision: String, productFilter: ProductFilter) throws -> [PackageContainerConstraint] {
+    func getDependencies(at revisionString: String, productFilter: ProductFilter) throws -> [PackageContainerConstraint] {
```
Comment (tomerd) [inline Sources/Workspace/ResolverPrecomputationProvider.swift:131]: please double check logic still looks right

File: Sources/Workspace/ResolverPrecomputationProvider.swift [hunk 2]
```diff
@@ -121,23 +122,25 @@ private struct LocalPackageContainer: PackageContainer {
 
     func getDependencies(at version: Version, productFilter: ProductFilter) throws -> [PackageContainerConstraint] {
         // Because of the implementation of `reversedVersions`, we should only get the exact same version.
-        precondition(dependency?.checkoutState?.version == version)
```
Comment (tomerd) [inline Sources/Workspace/ResolverPrecomputationProvider.swift:124]: please double check logic still looks right

File: Sources/Workspace/ResolverPrecomputationProvider.swift [hunk 3]
```diff
@@ -95,7 +96,7 @@ private struct LocalPackageContainer: PackageContainer {
     let currentToolsVersion: ToolsVersion
 
     func versionsAscending() throws -> [Version] {
-        if let version = dependency?.state.checkout?.version {
+        if case .version(let version, revision: _) = dependency?.state.checkout {
```
Comment (tomerd) [inline Sources/Workspace/ResolverPrecomputationProvider.swift:99]: please double check logic still looks right

File: Sources/Workspace/Workspace.swift [hunk 4]
```diff
@@ -2355,19 +2360,22 @@ extension Workspace {
...
                 // If we have a branch and we shouldn't be updating the
                 // branches, use the revision from pin instead (if present).
-                if branch != nil {
-                    if let pin = pinsStore.pins.first(where: { $0.packageRef == packageRef }),
-                        !updateBranches,
-                        pin.state.branch == branch {
-                        revision = pin.state.revision
+                if branch != nil, !updateBranches {
+                    if case .branch(branch, let pinRevision) = pinsStore.pins.first(where: { $0.packageRef == packageRef })?.state {
+                        revision = pinRevision
```
Comment (tomerd) [inline Sources/Workspace/Workspace.swift:2365]: please double check logic still looks right

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/PackageGraph/CheckoutState.swift [hunk 5]
```diff
@@ -15,50 +15,44 @@ import TSCUtility
...
-        self.branch = branch
-    }
-
-    /// Create a checkout state with given revision and branch.
-    public init(revision: Revision, branch: String? = nil) {
-        self.init(revision: revision, version: nil, branch: branch)
-    }
+    case revision(Revision)
+    case version(Version, revision: Revision)
+    case branch(String, revision: Revision)
```
Comment (neonichu) [inline Sources/PackageGraph/CheckoutState.swift:22]: Should we name the first parameter, too?

**Review by neonichu — APPROVED** [review]

File: Sources/PackageGraph/CheckoutState.swift (hunk 5 above)
Comment (tomerd) [reply Sources/PackageGraph/CheckoutState.swift:22]: I started with naming them but it was a bit awkward in the call site e.g. `let checkoutState: CheckoutState = .branch(branch: "", revision: "")`

Comment (abertelrud) [reply Sources/PackageGraph/CheckoutState.swift:22]: Right, the first parameter is often unnamed when the enum case makes it obvious, and I agree that that applies here.  So I think it should be unnamed.

File: Sources/Workspace/ResolverPrecomputationProvider.swift (hunk 3 above)
Comment (abertelrud) [reply Sources/Workspace/ResolverPrecomputationProvider.swift:99]: Looks correct to me.

File: Sources/Workspace/ResolverPrecomputationProvider.swift (hunk 2 above)
Comment (abertelrud) [reply Sources/Workspace/ResolverPrecomputationProvider.swift:124]: Looks correct to me.

File: Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift [hunk 6]
```diff
@@ -1054,7 +1054,13 @@ private final class PubGrubPackageContainer {
...
     /// Returns the pinned version for this package, if any.
     var pinnedVersion: Version? {
-        return self.pinsMap[self.underlying.package.identity]?.state.version
+        switch self.pinsMap[self.underlying.package.identity]?.state {
+        case .version(let version, _):
+            return version
+        default:
+            return .none
+        }
+        //return self.pinsMap[self.underlying.package.identity]?.state.version
```
Comment (abertelrud) [inline Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift:1063]: Was this intended to be left in?

**Review by abertelrud — APPROVED** [review]

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/PackageGraph/CheckoutState.swift (hunk 5 above)
Comment (tomerd) [reply Sources/PackageGraph/CheckoutState.swift]: ended up adding `name: ` to the branch variant

**Review by abertelrud — APPROVED** [review]

**Pull request merged**